# Activity 5.1 Text Similarity

Technological Institute of the Philippines | Quezon City - Computer Engineering
--- | ---
Course Code: | CPE 313
Code Title: | Advanced Machine Learning and Deep Learning
2nd Semester | AY 2025-2026
<u>**Hands-on Activity 5.1** | **Text Similarity**
**Name** | Adrian Justin Bulambao
**Section** | CPE32S3
**Date Performed**: | March 25, 2026
**Date Submitted**: | March 25, 2026
**Instructor**: | Engr. Neal Barton James Matira


#### Objective(s):

This activity aims to introduce how to implement text similarity

#### Intended Learning Outcomes (ILOs):
* Demonstrate how to remove special characters from text
* Demonstrate how to measure text simiarity using distance
* Demonstrate how to find text similarity using Count Vector and TF-IDF Vectorizer


#### Resources:
* Jupyter Notebook
* songdata.csv

#### Procedures
Load the necessary libraries

In [1]:
import nltk
import pandas as pd

In [2]:
data = pd.read_csv('songdata.csv')
data.head()

,artist,song,link,text
0,ABBA,Ahe's My Kind Of Girl,/a/abba/ahes+my+kind+of+girl_20598417.html,"Look at her face, it's a wonderful face \r\nA..."
1,ABBA,"Andante, Andante",/a/abba/andante+andante_20002708.html,"Take it easy with me, please \r\nTouch me gen..."
2,ABBA,As Good As New,/a/abba/as+good+as+new_20003033.html,I'll never know why I had to go \r\nWhy I had...
3,ABBA,Bang,/a/abba/bang_20598415.html,Making somebody happy is a question of give an...
4,ABBA,Bang-A-Boomerang,/a/abba/bang+a+boomerang_20002668.html,Making somebody happy is a question of give an...


Find the lyrics by the artist The Beatles

In [3]:
beatles = data[data.artist=='The Beatles']
beatles.head()

,artist,song,link,text
1198,The Beatles,A Shot Of Rhythm And Blues,/b/beatles/a+shot+of+rhythm+blues_20014867.html,"Well, if your hands start a-clappin' \r\nAnd ..."
1199,The Beatles,Across The Universe,/b/beatles/across+the+universe_10026507.html,Words are flowing out like \r\nEndless rain i...
1200,The Beatles,All I've Got To Do,/b/beatles/all+ive+got+to+do_10026646.html,"Whenever I want you around, yeah \r\nAll I go..."
1201,The Beatles,And I Love Her,/b/beatles/and+i+love+her_10026463.html,I give her all my love \r\nThat's all I do \...
1202,The Beatles,And Your Bird Can Sing,/b/beatles/and+your+bird+can+sing_10026364.html,You tell me that you've got everything you wan...


Check the number of Beatles songs in the data set


In [4]:
data.artist.value_counts().loc[['The Beatles']]

artist
The Beatles    178
Name: count, dtype: int64

Take a look at the first song's lyrics

In [5]:
beatles.iloc[0,3]

"Well, if your hands start a-clappin'  \r\nAnd your fingers start a-poppin'  \r\nAnd your feet start a-movin' around  \r\nAnd if you start to swing and sway  \r\n  \r\nWhen the band starts to play  \r\nA real cool way out sound  \r\nAnd if you get to can't help it and you can't sit down  \r\nYou feel like you gotta move around  \r\n  \r\nYou get a shot of rhythm and blues.  \r\nWith just a little rock and roll on the side  \r\nJust for good measure.  \r\nGet a pair of dancin' shoes  \r\n  \r\nWell, with your lover by your side  \r\nDon't you know you're gonna have a rockin' time, see'mon!  \r\nDon't you worry 'bout a thing  \r\nIf you start to dance and sing  \r\n  \r\nAnd chills come up on you  \r\nAnd if the rhythm finally gets you and the beat gets you too  \r\nWell, here's something for you to do  \r\n  \r\nGet a shot of rhythm and blues  \r\nWith just a little rock and roll on the side  \r\nJust for good measure  \r\nGet a pair of dancin' shoes  \r\n  \r\nWell, with your lover by 

Remove characters from the lyrics

In [6]:
beatles

,artist,song,link,text
1198,The Beatles,A Shot Of Rhythm And Blues,/b/beatles/a+shot+of+rhythm+blues_20014867.html,"Well, if your hands start a-clappin' \r\nAnd ..."
1199,The Beatles,Across The Universe,/b/beatles/across+the+universe_10026507.html,Words are flowing out like \r\nEndless rain i...
1200,The Beatles,All I've Got To Do,/b/beatles/all+ive+got+to+do_10026646.html,"Whenever I want you around, yeah \r\nAll I go..."
1201,The Beatles,And I Love Her,/b/beatles/and+i+love+her_10026463.html,I give her all my love \r\nThat's all I do \...
1202,The Beatles,And Your Bird Can Sing,/b/beatles/and+your+bird+can+sing_10026364.html,You tell me that you've got everything you wan...
...,...,...,...,...
24826,The Beatles,Nowhere Man,/b/beatles/nowhere+man_10026490.html,He's a real nowhere man \r\nSitting in his no...
24827,The Beatles,"Ob-La-Di, Ob-La-Da",/b/beatles/ob+la+di+ob+la+da_10026611.html,Desmond has a barrow in the marketplace \r\nM...
24828,The Beatles,Octopus's Garden,/b/beatles/octopuss+garden_10026529.html,I'd like to be under the sea \r\nIn an octopu...
24829,The Beatles,Oh! Darling,/b/beatles/oh+darling_10026075.html,"Oh! Darling, please believe me \r\nI'll never..."


In [7]:
import re

newline = lambda x: re.sub('\n', ' ', x) # remove \n
alphanumeric = lambda x: re.sub('\w*\d\w*', ' ', x) # remove alphanumeric words

corpus = beatles.text.map(newline).map(alphanumeric)
corpus.head()

<>:4: SyntaxWarning: invalid escape sequence '\w'
<>:4: SyntaxWarning: invalid escape sequence '\w'
C:\Users\JUSTIN\AppData\Local\Temp\ipykernel_19196\2279837142.py:4: SyntaxWarning: invalid escape sequence '\w'
  alphanumeric = lambda x: re.sub('\w*\d\w*', ' ', x) # remove alphanumeric words


1198    Well, if your hands start a-clappin'  \r And y...
1199    Words are flowing out like  \r Endless rain in...
1200    Whenever I want you around, yeah  \r All I got...
1201    I give her all my love  \r That's all I do  \r...
1202    You tell me that you've got everything you wan...
Name: text, dtype: str

Create a document term matrix using Count Vectorizer with the stop words turned on to English

In [8]:
from sklearn.feature_extraction.text import CountVectorizer
          
cv = CountVectorizer(stop_words="english")
X = cv.fit_transform(corpus).toarray()

dt = pd.DataFrame(X, columns=cv.get_feature_names_out()).set_index(beatles.song)
dt.head()

,aaahhh,aah,abc,aches,aching,acquainted,act,actors,acts,add,...,yes,yesterday,yoko,young,younger,youre,youu,zealand,zoo,zu
song,,,,,,,,,,,,,,,,,,,,,
A Shot Of Rhythm And Blues,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
Across The Universe,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
All I've Got To Do,0,0,0,0,0,0,0,0,0,0,...,2,0,0,0,0,0,0,0,0,0
And I Love Her,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
And Your Bird Can Sing,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


Set display options so that the lyrics aren't cut off

In [9]:
pd.set_option('display.max_colwidth', None)

In [10]:
beatles[beatles.song=='Imagine'].text

24783    Imagine there's no heaven  \r\nIt's easy if you try  \r\nNo hell below us  \r\nAbove us only sky  \r\nImagine all the people  \r\nLiving for today...  \r\n  \r\nImagine there's no countries  \r\nIt isn't hard to do  \r\nNothing to kill or die for  \r\nAnd no religion too  \r\nImagine all the people  \r\nLiving life in peace...  \r\n  \r\nYou may say I'm a dreamer  \r\nBut I'm not the only one  \r\nI hope someday you'll join us  \r\nAnd the world will be as one  \r\n  \r\nImagine no possessions  \r\nI wonder if you can  \r\nNo need for greed or hunger  \r\nA brotherhood of man  \r\nImagine all the people  \r\nSharing all the world...  \r\n  \r\nYou may say I'm a dreamer  \r\nBut I'm not the only one  \r\nI hope someday you'll join us  \r\nAnd the world will live as one\r\n\r\n
Name: text, dtype: str

Imagine lyrics in Count Vectorizer form


In [11]:
imagine = list(dt.loc['Imagine'])
imagine[:20]

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

Define the cosine similarity calculation

In [12]:
from numpy import dot
from numpy.linalg import norm

cosine = lambda v1, v2: dot(v1, v2) / (norm(v1) * norm(v2))

Calculate all similarities and sort by the most similar

In [13]:
similarity = [cosine(imagine, song) for song in X]
sorted(list(zip(similarity, beatles.song)), reverse=True)[1:6]

[(np.float64(0.22757944185316942), "I'll Cry Instead"),
 (np.float64(0.1738270040532526), 'In My Life'),
 (np.float64(0.16574838603294895), 'Eleanor Rigby'),
 (np.float64(0.1490280831849844), 'In Spite Of All The Danger'),
 (np.float64(0.14592261028699943), 'All My Loving')]

Create the TF-IDF Vectorizer

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer
          
cv_tfidf = TfidfVectorizer(stop_words="english")
X_tfidf = cv_tfidf.fit_transform(corpus).toarray()

dt_tfidf = pd.DataFrame(X_tfidf, columns=cv_tfidf.get_feature_names_out()).set_index(beatles.song)
dt_tfidf.head()

,aaahhh,aah,abc,aches,aching,acquainted,act,actors,acts,add,...,yes,yesterday,yoko,young,younger,youre,youu,zealand,zoo,zu
song,,,,,,,,,,,,,,,,,,,,,
A Shot Of Rhythm And Blues,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
Across The Universe,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
All I've Got To Do,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.120483,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
And I Love Her,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
And Your Bird Can Sing,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer
          
cv_tfidf = TfidfVectorizer(stop_words="english")
X_tfidf = cv_tfidf.fit_transform(corpus).toarray()

dt_tfidf = pd.DataFrame(X_tfidf, columns=cv_tfidf.get_feature_names_out()).set_index(beatles.song)
dt_tfidf.head()

,aaahhh,aah,abc,aches,aching,acquainted,act,actors,acts,add,...,yes,yesterday,yoko,young,younger,youre,youu,zealand,zoo,zu
song,,,,,,,,,,,,,,,,,,,,,
A Shot Of Rhythm And Blues,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
Across The Universe,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
All I've Got To Do,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.120483,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
And I Love Her,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
And Your Bird Can Sing,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


Calculate all similarities and sort by the most similar using the TF-IDF data

In [16]:
similarity_tfidf = [cosine(imagine, song) for song in X_tfidf]
sorted(list(zip(similarity_tfidf, beatles.song)), reverse=True)[1:6]

[(np.float64(0.19266277925834221), "I'll Cry Instead"),
 (np.float64(0.18004423848592196), "I'll Get You"),
 (np.float64(0.1551371744116933), 'Eleanor Rigby'),
 (np.float64(0.15054395512050195), 'In My Life'),
 (np.float64(0.12306875209548622), 'Love Of The Loved')]

Compare the lyrics of Imagine and I'll Cry Instead

In [17]:
# Imagine lyrics
beatles[beatles.song=='Imagine'].text

24783    Imagine there's no heaven  \r\nIt's easy if you try  \r\nNo hell below us  \r\nAbove us only sky  \r\nImagine all the people  \r\nLiving for today...  \r\n  \r\nImagine there's no countries  \r\nIt isn't hard to do  \r\nNothing to kill or die for  \r\nAnd no religion too  \r\nImagine all the people  \r\nLiving life in peace...  \r\n  \r\nYou may say I'm a dreamer  \r\nBut I'm not the only one  \r\nI hope someday you'll join us  \r\nAnd the world will be as one  \r\n  \r\nImagine no possessions  \r\nI wonder if you can  \r\nNo need for greed or hunger  \r\nA brotherhood of man  \r\nImagine all the people  \r\nSharing all the world...  \r\n  \r\nYou may say I'm a dreamer  \r\nBut I'm not the only one  \r\nI hope someday you'll join us  \r\nAnd the world will live as one\r\n\r\n
Name: text, dtype: str

In [18]:
beatles[beatles.song=='I\'ll Cry Instead'].text

24774    I've got every reason on earth to be mad  \r\n'Cause I just lost the only girl I had  \r\nIf I could get my way  \r\nI'd get myself locked up today  \r\nBut I can't, so I'll cry instead  \r\n  \r\nI've got a chip on my shoulder that's bigger that my feet  \r\nI can't talk to people that I meet  \r\nIf I could see you now  \r\nI'd try to make you sad somehow  \r\nBut I can't, so I'll cry instead  \r\n  \r\nDon't want to cry when there's people there  \r\nI get shy when they start to stare  \r\nI'm gonna hide myself away  \r\nBut I'll come back again someday  \r\n  \r\nAnd when you do you'd better hide all the girls  \r\nI'm gonna break their hearts all round the world  \r\nYes, I'm gonna break them in two  \r\nAnd show you what your lovin' man can do  \r\nUntil then I'll cry instead\r\n\r\n
Name: text, dtype: str

**Compare the two lyrics based on the output**

In [19]:
be1 = dt.loc['Imagine'].values
be2 = dt.loc["I'll Cry Instead"].values
similarity_songs = [cosine(be1, be2)]
print(similarity_songs)

[np.float64(0.22757944185316942)]


Answer:
- The two songs have a 22.76% overlap in words used in the words. this ignores the order and frequency in the songs.

Find the two Beatles songs that are the most similar
* Using Count Vectorizer
* Using TF-IDF Vectorizer

Calculate the cosine similarity between all combinations of documents

In [20]:
from itertools import combinations

pairs = list(combinations(beatles.song.index, 2)) # all song index combos
pairs_0 = list(combinations(range(len(beatles)), 2)) # all index combos starting with (0,1)
song_pairs = [(beatles.song[a_index], beatles.song[b_index]) for (a_index, b_index) in pairs]
song_pairs[:5]

[('A Shot Of Rhythm And Blues', 'Across The Universe'),
 ('A Shot Of Rhythm And Blues', "All I've Got To Do"),
 ('A Shot Of Rhythm And Blues', 'And I Love Her'),
 ('A Shot Of Rhythm And Blues', 'And Your Bird Can Sing'),
 ('A Shot Of Rhythm And Blues', 'Another Girl')]

 Calculate the cosine similarity of the Count Vectorizer document-term matrix

In [21]:
results = [cosine(X[a_index], X[b_index]) for (a_index, b_index) in pairs_0]
sorted(zip(results, song_pairs), reverse=True)[:5]

[(np.float64(0.8548876342047523), ('All You Need Is Love', 'Love Me Do')),
 (np.float64(0.8359173188630292),
  ('If I Needed Someone', 'If I Needed Someone To Love')),
 (np.float64(0.8058125835571367), ('And I Love Her', 'Love Me Do')),
 (np.float64(0.7930110315846619), ('If I Fell', 'Love Me Do')),
 (np.float64(0.7478391382739419), ('And I Love Her', 'All You Need Is Love'))]

Calculate the cosine similarity of the TF-IDF Vectorizer document-term matrix

In [22]:
results = [cosine(X_tfidf[a_index], X_tfidf[b_index]) for (a_index, b_index) in pairs_0]
sorted(zip(results, song_pairs), reverse=True)[:5]

[(np.float64(0.8717047783261523),
  ('If I Needed Someone', 'If I Needed Someone To Love')),
 (np.float64(0.6599353568343105), ('All You Need Is Love', 'Love Me Do')),
 (np.float64(0.6502842517883831), ('Have A Banana!', 'Crinsk Dee Night')),
 (np.float64(0.6232797327325461), ("Don't Let Me Down", 'Let It Be')),
 (np.float64(0.5275364336322863), ('If I Fell', 'Love Me Do'))]

In [23]:
 beatles[beatles.song=='All You Need Is Love'].text

24695    Love, love, love, love, love, love, love, love, love.  \r\n  \r\nThere's nothing you can do that can't be done.  \r\nNothing you can sing that can't be sung.  \r\nNothing you can say, but you can learn  \r\nHow to play the game  \r\nIt's easy.  \r\nNothing you can make that can't be made.  \r\nNo one you can save that can't be saved.  \r\nNothing you can do, but you can learn  \r\nHow to be you in time  \r\nIt's easy.  \r\n  \r\nAll you need is love, all you need is love,  \r\nAll you need is love, love. Love is all you need.  \r\nLove, love, love, love, love, love, love, love, love.  \r\nAll you need is love, all you need is love,  \r\nAll you need is love, love. Love is all you need.  \r\n  \r\nThere's nothing you can know that isn't known.  \r\nNothing you can see that isn't shown.  \r\nThere's nowhere you can be that isn't where  \r\nYou're meant to be  \r\nIt's easy.  \r\n  \r\nAll you need is love, all you need is love,  \r\nAll you need is love, love. Love is all you ne

In [24]:
beatles[beatles.song=='Love Me Do'].text

24805    Love, love me do.  \r\nYou know I love you,  \r\nI'll always be true,  \r\nSo please, love me do.  \r\nWhoa, love me do.  \r\n  \r\nLove, love me do.  \r\nYou know I love you,  \r\nI'll always be true,  \r\nSo please, love me do.  \r\nWhoa, love me do.  \r\n  \r\nSomeone to love,  \r\nSomebody new.  \r\nSomeone to love,  \r\nSomeone like you.  \r\n  \r\nLove, love me do.  \r\nYou know I love you,  \r\nI'll always be true,  \r\nSo please, love me do.  \r\nWhoa, love me do.  \r\n  \r\nLove, love me do.  \r\nYou know I love you,  \r\nI'll always be true,  \r\nSo please, love me do.  \r\nWhoa, love me do.  \r\nYeah, love me do.  \r\nWhoa, oh, love me do.\r\n\r\n
Name: text, dtype: str

In [25]:
beatles[beatles.song=='If I Needed Someone'].text


24768    If I needed someone to love  \r\nYou're the one that I'd be thinking of  \r\nIf I needed someone  \r\n  \r\nIf I had some more time to spend  \r\nThen I guess I'd be with you my friend  \r\nIf I needed someone  \r\nHad you come some other day  \r\nThen it might not have been like this  \r\nBut you see now I'm too much in love  \r\n  \r\nCarve your number on my wall  \r\nAnd maybe you will get a call from me  \r\nIf I needed someone  \r\nAh, ah, ah, ah  \r\n  \r\nIf I had some more time to spend  \r\nThen I guess I'd be with you my friend  \r\nIf I needed someone  \r\nHad you come some other day  \r\nThen it might not have been like this  \r\nBut you see now I'm too much in love  \r\n  \r\nCarve your number on my wall  \r\nAnd maybe you will get a call from me  \r\nIf I needed someone  \r\nAh, ah\r\n\r\n
Name: text, dtype: str

In [26]:
beatles[beatles.song=='If I Needed Someone To Love'].text

24769    If I needed someone to love  \r\nYoure the woman I'd be thinking of  \r\nIf I needed someone  \r\n  \r\nIf I had some more time to spend  \r\nThen I guess I'd be with you my friend  \r\nIf I needed someone  \r\n  \r\nHad you come some other day  \r\nThen it might not have been like this  \r\nBut you see now I'm too much in love  \r\n  \r\nCarve your number on my wall  \r\nAnd maybe you will get a call from me  \r\nIf I needed someone  \r\n  \r\nIf I had some more time to spend  \r\nThen I guess I'd be with you my friend  \r\nIf I needed someone  \r\n  \r\nHad you come some other day  \r\nThen it might not have been like this  \r\nBut you see now I'm too much in love  \r\n  \r\nCarve your number on my wall  \r\nAnd maybe you will get a call from me  \r\nIf I needed someone  \r\nIf I needed someone to love  \r\nYoure the woman I'd be thinking of  \r\nIf I needed someone  \r\n  \r\nIf I had some more time to spend  \r\nThen I guess I'd be with you my friend  \r\nIf I needed someo

In [27]:
beatles[beatles.song=='Have A Banana!'].text

1221    [Speech]  \r\n  \r\nBrian Matthew: Is that it? Is that the end?  \r\nPaul: Yeah, yeah, that's it.  \r\nJohn: Fade, fade!  \r\nBrian: Good track. Oh, well, we'll stop there, stop there, stop there.  \r\nJohn: What an end!  \r\nBrian: Quiet! All right, George.  \r\nJohn: Fade!  \r\nBrian: Hold it!  \r\nGeorge: Oh, thank you.  \r\nJohn: Fade, you silly.  \r\nBrian: Well, we did. We did that. Oh, no! No! We've done that bit!  \r\nJohn: The train comes in now.  \r\nBrian: We did that.  \r\nJohn: Yeah.  \r\nBrian: To pove we weren't playing the record, then, you see. 'Cause,\r\notherwise, there's no point in you being here, is there? Ha, ha, ha!  \r\nJohn: Yeah, we did that, 'cause it sounds just like it, don't it?  \r\nBrian: Pretty cool lot of fellows, aren't you? Here, Ringo, have a banana,\r\ncatch!\r\n\r\n
Name: text, dtype: str

In [28]:
beatles[beatles.song=='Crinsk Dee Night'].text

24713    [Speech]  \r\n  \r\nBrian Matthew: The next few minutes, we're in the lap of the gods and the\r\nhands of the Beatles. In my young days, when I was a lad, they used to have\r\nactors in films and now that they--  \r\nPaul: Yes?  \r\nJohn: Hey! Listen!  \r\nPaul: It's all changed, now, Brian. They're not doing that, no actors.  \r\nJohn: It's all changed, now.  \r\nBrian: But this is what I wonder. In those days, the actors used to say their\r\nbest bits were left on the cutting room floor. Did you find that?  \r\nJohn: No, no, no, those were the good bits in the film. You should have seen\r\nthe rest.  \r\nBrian: Yes?  \r\nJohn: Rubbish!  \r\nBrian: Was it, really?  \r\nJohn: Even worse, yes.  \r\nBrian: Who was worst?  \r\nJohn: Oh, Paul.  \r\nBrian: I see.  \r\nPaul: I think John was about the worst.  \r\nJohn: No, it was you.  \r\nPaul: Oh, Ringo was very good. He was. He's a good lad.  \r\nBrian: He was. They're saying he's a new Charlie Chaplin. Do you think that's\r\nrig

Compare the results. Which Vectorizer seems to do a better job?

Answer:
- The TF-IDF vectorizer did a better job than the Count Vector, as based on the results, the TF-IDF has a better score than the vector count in terms of calculating the cosine similarity.

<hr>
<br>
<br>
<br>

#### **Supplementary Activity**

* Use a dataset that consists of text document
* Remove special characters
* Compare the similar document using Count Vector and TF-IDF Vectorizer
* Use the cosine similarity to measure the similarity of text

The dataset used for the supplementary is a poem dataset from kaggle, taken from this link: https://www.kaggle.com/datasets/mexwell/poem-dataset , each poem has a class, and a total of 6 classes.

In [41]:
import pandas as pd
import nltk
from nltk.tokenize import TreebankWordTokenizer
from nltk.corpus import stopwords
from nltk.tag import pos_tag
import re

# loading the dataset
poem = pd.read_csv("final_df_emotions(remove-bias).csv")
poem['title'] = ["Poem " + str(i+1) for i in range(len(poem))]
poem = poem [['label', 'title', 'poem content']]
poem.head()


,label,title,poem content
0,sadness,Poem 1,"Let the bird of loudest lay\r\nOn the sole Arabian tree\r\nHerald sad and trumpet be,\r\nTo whose sound chaste wings obey.\r\n\r\nBut thou shrieking harbinger,\r\nFoul precurrer of the fiend,\r\nAugur of the fever's end,\r\nTo this troop come thou not near.\r\n\r\nFrom this session interdict\r\nEvery fowl of tyrant wing,\r\nSave the eagle, feather'd king;\r\nKeep the obsequy so strict.\r\n\r\nLet the priest in surplice white,\r\nThat defunctive music can,\r\nBe the death-divining swan,\r\nLest the requiem lack his right.\r\n\r\nAnd thou treble-dated crow,\r\nThat thy sable gender mak'st\r\nWith the breath thou giv'st and tak'st,\r\n'Mongst our mourners shalt thou go.\r\n\r\nHere the anthem doth commence:\r\nLove and constancy is dead;\r\nPhoenix and the Turtle fled\r\nIn a mutual flame from hence.\r\n\r\nSo they lov'd, as love in twain\r\nHad the essence but in one;\r\nTwo distincts, division none:\r\nNumber there in love was slain.\r\n\r\nHearts remote, yet not asunder;\r\nDistance and no space was seen\r\n'Twixt this Turtle and his queen:\r\nBut in them it were a wonder.\r\n\r\nSo between them love did shine\r\nThat the Turtle saw his right\r\nFlaming in the Phoenix' sight:\r\nEither was the other's mine.\r\n\r\nProperty was thus appalled\r\nThat the self was not the same;\r\nSingle nature's double name\r\nNeither two nor one was called.\r\n\r\nReason, in itself confounded,\r\n"
1,anger,Poem 2,"Sir Charles into my chamber coming in,\r\nWhen I was writing of my Fairy Queen;\r\nI praysaid hewhen Queen Mab you do see\r\nPresent my service to her Majesty:\r\nAnd tell her I have heard Fame's loud report\r\nBoth of her beauty and her stately court.\r\nWhen I Queen Mab within my fancy viewed,\r\nMy thoughts bowed low, fearing I should be rude;\r\nKissing her garment thin which fancy made,\r\nI knelt upon a thought, like one that prayed;\r\nAnd then, in whispers soft, I did present\r\nHis humble service which in mirth was sent;\r\nThus by imagination I have been\r\nIn Fairy court and seen the Fairy Queen."
2,anger,Poem 3,"Our vice runs beyond all that old men saw,\r\nAnd far authentically above our laws,\r\nAnd scorning virtues safe and golden mean,\r\nSits uncontrolled upon the high extreme.\r\nCirces, thy monsters painted out the hue,\r\nOf feigned filthiness, but ours is true.\r\nOur vice puts down all proverbs and all themes,\r\nOur vice excels all fables and all dreams."
3,anger,Poem 4,"Lo I the man, whose Muse whilome did maske,\r\nAs time her taught in lowly Shepheards weeds,\r\nAm now enforst a far unfitter taske,\r\nFor trumpets sterne to chaunge mine Oaten reeds,\r\nAnd sing of Knights and Ladies gentle deeds;\r\nWhose prayses having slept in silence long,\r\nMe, all too meane, the sacred Muse areeds\r\nTo blazon broad emongst her learned throng:\r\nFierce warres and faithful loves shall moralize my song.\r\nHelpe then, O holy Virgin chiefe of nine,\r\nThy weaker Novice to performe thy will,\r\nLay forth out of thine everlasting scryne\r\nThe antique rolles, which there lye hidden still,\r\nOf Faerie knights and fairest Tanaquill,\r\nWhom that most noble Briton Prince so long\r\nSought through the world, and suffered so much ill,\r\nThat I must rue his undeserved wrong:\r\nO helpe thou my weake wit, and sharpen my dull tong.\r\nAnd thou most dreaded impe of highest Jove,\r\nFaire Venus sonne, that with thy cruell dart\r\nAt that good knight so cunningly didst rove,\r\nThat glorious fire it kindled in his hart,\r\nLay now thy deadly Heben bow apart,\r\nAnd with thy mother milde come to mine ayde:\r\nCome both, and with you bring triumphant Mart,\r\nIn loves and gentle jollities arrayd,\r\nAfter his murdrous spoiles and bloudy rage allayd.\r\nAnd with them eke, O Goddesse heavenly bright,\r\nMirrour of grace and M"
4,sadness,Poem 5,"Long have I longd to see my love againe,\r\nStill have I wisht, but never could obtaine it;\r\nRather than all the world (if I might gaine it)\r\nWould

* Tokenize (words)

In [42]:
nltk.download('stopwords')
nltk.download('averaged_perceptron_tagger')
nltk.download('averaged_perceptron_tagger_eng')
stop_words = set(stopwords.words('english'))
tokenizer = TreebankWordTokenizer()

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\JUSTIN\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\JUSTIN\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     C:\Users\JUSTIN\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger_eng is already up-to-
[nltk_data]       date!


In [43]:
poem['tokens_basic'] = poem['poem content'].map(lambda x: tokenizer.tokenize(x))
poem.head(1)

,label,title,poem content,tokens_basic
0,sadness,Poem 1,"Let the bird of loudest lay\r\nOn the sole Arabian tree\r\nHerald sad and trumpet be,\r\nTo whose sound chaste wings obey.\r\n\r\nBut thou shrieking harbinger,\r\nFoul precurrer of the fiend,\r\nAugur of the fever's end,\r\nTo this troop come thou not near.\r\n\r\nFrom this session interdict\r\nEvery fowl of tyrant wing,\r\nSave the eagle, feather'd king;\r\nKeep the obsequy so strict.\r\n\r\nLet the priest in surplice white,\r\nThat defunctive music can,\r\nBe the death-divining swan,\r\nLest the requiem lack his right.\r\n\r\nAnd thou treble-dated crow,\r\nThat thy sable gender mak'st\r\nWith the breath thou giv'st and tak'st,\r\n'Mongst our mourners shalt thou go.\r\n\r\nHere the anthem doth commence:\r\nLove and constancy is dead;\r\nPhoenix and the Turtle fled\r\nIn a mutual flame from hence.\r\n\r\nSo they lov'd, as love in twain\r\nHad the essence but in one;\r\nTwo distincts, division none:\r\nNumber there in love was slain.\r\n\r\nHearts remote, yet not asunder;\r\nDistance and no space was seen\r\n'Twixt this Turtle and his queen:\r\nBut in them it were a wonder.\r\n\r\nSo between them love did shine\r\nThat the Turtle saw his right\r\nFlaming in the Phoenix' sight:\r\nEither was the other's mine.\r\n\r\nProperty was thus appalled\r\nThat the self was not the same;\r\nSingle nature's double name\r\nNeither two nor one was called.\r\n\r\nReason, in itself confounded,\r\n","[Let, the, bird, of, loudest, lay, On, the, sole, Arabian, tree, Herald, sad, and, trumpet, be, ,, To, whose, sound, chaste, wings, obey., But, thou, shrieking, harbinger, ,, Foul, precurrer, of, the, fiend, ,, Augur, of, the, fever, 's, end, ,, To, this, troop, come, thou, not, near., From, this, session, interdict, Every, fowl, of, tyrant, wing, ,, Save, the, eagle, ,, feather, 'd, king, ;, Keep, the, obsequy, so, strict., Let, the, priest, in, surplice, white, ,, That, defunctive, music, can, ,, Be, the, death-divining, swan, ,, Lest, the, requiem, lack, his, right., And, thou, treble-dated, crow, ,, That, ...]"


* Tokenize (words with remove punctuation) / Tokenize (words with remove special characters and numbers)

In [44]:
clean_text = lambda x: re.sub(r'[^A-Za-z\s]', ' ', x)
poem['tokens_clean'] = poem['poem content'].map(clean_text).map(lambda x: x.split())
poem.head(1)

,label,title,poem content,tokens_basic,tokens_clean
0,sadness,Poem 1,"Let the bird of loudest lay\r\nOn the sole Arabian tree\r\nHerald sad and trumpet be,\r\nTo whose sound chaste wings obey.\r\n\r\nBut thou shrieking harbinger,\r\nFoul precurrer of the fiend,\r\nAugur of the fever's end,\r\nTo this troop come thou not near.\r\n\r\nFrom this session interdict\r\nEvery fowl of tyrant wing,\r\nSave the eagle, feather'd king;\r\nKeep the obsequy so strict.\r\n\r\nLet the priest in surplice white,\r\nThat defunctive music can,\r\nBe the death-divining swan,\r\nLest the requiem lack his right.\r\n\r\nAnd thou treble-dated crow,\r\nThat thy sable gender mak'st\r\nWith the breath thou giv'st and tak'st,\r\n'Mongst our mourners shalt thou go.\r\n\r\nHere the anthem doth commence:\r\nLove and constancy is dead;\r\nPhoenix and the Turtle fled\r\nIn a mutual flame from hence.\r\n\r\nSo they lov'd, as love in twain\r\nHad the essence but in one;\r\nTwo distincts, division none:\r\nNumber there in love was slain.\r\n\r\nHearts remote, yet not asunder;\r\nDistance and no space was seen\r\n'Twixt this Turtle and his queen:\r\nBut in them it were a wonder.\r\n\r\nSo between them love did shine\r\nThat the Turtle saw his right\r\nFlaming in the Phoenix' sight:\r\nEither was the other's mine.\r\n\r\nProperty was thus appalled\r\nThat the self was not the same;\r\nSingle nature's double name\r\nNeither two nor one was called.\r\n\r\nReason, in itself confounded,\r\n","[Let, the, bird, of, loudest, lay, On, the, sole, Arabian, tree, Herald, sad, and, trumpet, be, ,, To, whose, sound, chaste, wings, obey., But, thou, shrieking, harbinger, ,, Foul, precurrer, of, the, fiend, ,, Augur, of, the, fever, 's, end, ,, To, this, troop, come, thou, not, near., From, this, session, interdict, Every, fowl, of, tyrant, wing, ,, Save, the, eagle, ,, feather, 'd, king, ;, Keep, the, obsequy, so, strict., Let, the, priest, in, surplice, white, ,, That, defunctive, music, can, ,, Be, the, death-divining, swan, ,, Lest, the, requiem, lack, his, right., And, thou, treble-dated, crow, ,, That, ...]","[Let, the, bird, of, loudest, lay, On, the, sole, Arabian, tree, Herald, sad, and, trumpet, be, To, whose, sound, chaste, wings, obey, But, thou, shrieking, harbinger, Foul, precurrer, of, the, fiend, Augur, of, the, fever, s, end, To, this, troop, come, thou, not, near, From, this, session, interdict, Every, fowl, of, tyrant, wing, Save, the, eagle, feather, d, king, Keep, the, obsequy, so, strict, Let, the, priest, in, surplice, white, That, defunctive, music, can, Be, the, death, divining, swan, Lest, the, requiem, lack, his, right, And, thou, treble, dated, crow, That, thy, sable, gender, mak, st, With, the, breath, thou, ...]"


* Tokenize (words with remove punctuation and all lowercase)

In [45]:
poem['tokens_lower'] = poem['poem content'].map(clean_text).map(lambda x: x.lower().split())
poem.head(1)


,label,title,poem content,tokens_basic,tokens_clean,tokens_lower
0,sadness,Poem 1,"Let the bird of loudest lay\r\nOn the sole Arabian tree\r\nHerald sad and trumpet be,\r\nTo whose sound chaste wings obey.\r\n\r\nBut thou shrieking harbinger,\r\nFoul precurrer of the fiend,\r\nAugur of the fever's end,\r\nTo this troop come thou not near.\r\n\r\nFrom this session interdict\r\nEvery fowl of tyrant wing,\r\nSave the eagle, feather'd king;\r\nKeep the obsequy so strict.\r\n\r\nLet the priest in surplice white,\r\nThat defunctive music can,\r\nBe the death-divining swan,\r\nLest the requiem lack his right.\r\n\r\nAnd thou treble-dated crow,\r\nThat thy sable gender mak'st\r\nWith the breath thou giv'st and tak'st,\r\n'Mongst our mourners shalt thou go.\r\n\r\nHere the anthem doth commence:\r\nLove and constancy is dead;\r\nPhoenix and the Turtle fled\r\nIn a mutual flame from hence.\r\n\r\nSo they lov'd, as love in twain\r\nHad the essence but in one;\r\nTwo distincts, division none:\r\nNumber there in love was slain.\r\n\r\nHearts remote, yet not asunder;\r\nDistance and no space was seen\r\n'Twixt this Turtle and his queen:\r\nBut in them it were a wonder.\r\n\r\nSo between them love did shine\r\nThat the Turtle saw his right\r\nFlaming in the Phoenix' sight:\r\nEither was the other's mine.\r\n\r\nProperty was thus appalled\r\nThat the self was not the same;\r\nSingle nature's double name\r\nNeither two nor one was called.\r\n\r\nReason, in itself confounded,\r\n","[Let, the, bird, of, loudest, lay, On, the, sole, Arabian, tree, Herald, sad, and, trumpet, be, ,, To, whose, sound, chaste, wings, obey., But, thou, shrieking, harbinger, ,, Foul, precurrer, of, the, fiend, ,, Augur, of, the, fever, 's, end, ,, To, this, troop, come, thou, not, near., From, this, session, interdict, Every, fowl, of, tyrant, wing, ,, Save, the, eagle, ,, feather, 'd, king, ;, Keep, the, obsequy, so, strict., Let, the, priest, in, surplice, white, ,, That, defunctive, music, can, ,, Be, the, death-divining, swan, ,, Lest, the, requiem, lack, his, right., And, thou, treble-dated, crow, ,, That, ...]","[Let, the, bird, of, loudest, lay, On, the, sole, Arabian, tree, Herald, sad, and, trumpet, be, To, whose, sound, chaste, wings, obey, But, thou, shrieking, harbinger, Foul, precurrer, of, the, fiend, Augur, of, the, fever, s, end, To, this, troop, come, thou, not, near, From, this, session, interdict, Every, fowl, of, tyrant, wing, Save, the, eagle, feather, d, king, Keep, the, obsequy, so, strict, Let, the, priest, in, surplice, white, That, defunctive, music, can, Be, the, death, divining, swan, Lest, the, requiem, lack, his, right, And, thou, treble, dated, crow, That, thy, sable, gender, mak, st, With, the, breath, thou, ...]","[let, the, bird, of, loudest, lay, on, the, sole, arabian, tree, herald, sad, and, trumpet, be, to, whose, sound, chaste, wings, obey, but, thou, shrieking, harbinger, foul, precurrer, of, the, fiend, augur, of, the, fever, s, end, to, this, troop, come, thou, not, near, from, this, session, interdict, every, fowl, of, tyrant, wing, save, the, eagle, feather, d, king, keep, the, obsequy, so, strict, let, the, priest, in, surplice, white, that, defunctive, music, can, be, the, death, divining, swan, lest, the, requiem, lack, his, right, and, thou, treble, dated, crow, that, thy, sable, gender, mak, st, with, the, breath, thou, ...]"


* Tokenize (words with remove punctuation, all lowercase and remove stop words)


In [46]:
poem['tokens_no_stop'] = poem['poem content'].map(clean_text).map(lambda x: x.lower().split()).map(lambda words: [w for w in words if w not in stop_words])
poem.head(1)

,label,title,poem content,tokens_basic,tokens_clean,tokens_lower,tokens_no_stop
0,sadness,Poem 1,"Let the bird of loudest lay\r\nOn the sole Arabian tree\r\nHerald sad and trumpet be,\r\nTo whose sound chaste wings obey.\r\n\r\nBut thou shrieking harbinger,\r\nFoul precurrer of the fiend,\r\nAugur of the fever's end,\r\nTo this troop come thou not near.\r\n\r\nFrom this session interdict\r\nEvery fowl of tyrant wing,\r\nSave the eagle, feather'd king;\r\nKeep the obsequy so strict.\r\n\r\nLet the priest in surplice white,\r\nThat defunctive music can,\r\nBe the death-divining swan,\r\nLest the requiem lack his right.\r\n\r\nAnd thou treble-dated crow,\r\nThat thy sable gender mak'st\r\nWith the breath thou giv'st and tak'st,\r\n'Mongst our mourners shalt thou go.\r\n\r\nHere the anthem doth commence:\r\nLove and constancy is dead;\r\nPhoenix and the Turtle fled\r\nIn a mutual flame from hence.\r\n\r\nSo they lov'd, as love in twain\r\nHad the essence but in one;\r\nTwo distincts, division none:\r\nNumber there in love was slain.\r\n\r\nHearts remote, yet not asunder;\r\nDistance and no space was seen\r\n'Twixt this Turtle and his queen:\r\nBut in them it were a wonder.\r\n\r\nSo between them love did shine\r\nThat the Turtle saw his right\r\nFlaming in the Phoenix' sight:\r\nEither was the other's mine.\r\n\r\nProperty was thus appalled\r\nThat the self was not the same;\r\nSingle nature's double name\r\nNeither two nor one was called.\r\n\r\nReason, in itself confounded,\r\n","[Let, the, bird, of, loudest, lay, On, the, sole, Arabian, tree, Herald, sad, and, trumpet, be, ,, To, whose, sound, chaste, wings, obey., But, thou, shrieking, harbinger, ,, Foul, precurrer, of, the, fiend, ,, Augur, of, the, fever, 's, end, ,, To, this, troop, come, thou, not, near., From, this, session, interdict, Every, fowl, of, tyrant, wing, ,, Save, the, eagle, ,, feather, 'd, king, ;, Keep, the, obsequy, so, strict., Let, the, priest, in, surplice, white, ,, That, defunctive, music, can, ,, Be, the, death-divining, swan, ,, Lest, the, requiem, lack, his, right., And, thou, treble-dated, crow, ,, That, ...]","[Let, the, bird, of, loudest, lay, On, the, sole, Arabian, tree, Herald, sad, and, trumpet, be, To, whose, sound, chaste, wings, obey, But, thou, shrieking, harbinger, Foul, precurrer, of, the, fiend, Augur, of, the, fever, s, end, To, this, troop, come, thou, not, near, From, this, session, interdict, Every, fowl, of, tyrant, wing, Save, the, eagle, feather, d, king, Keep, the, obsequy, so, strict, Let, the, priest, in, surplice, white, That, defunctive, music, can, Be, the, death, divining, swan, Lest, the, requiem, lack, his, right, And, thou, treble, dated, crow, That, thy, sable, gender, mak, st, With, the, breath, thou, ...]","[let, the, bird, of, loudest, lay, on, the, sole, arabian, tree, herald, sad, and, trumpet, be, to, whose, sound, chaste, wings, obey, but, thou, shrieking, harbinger, foul, precurrer, of, the, fiend, augur, of, the, fever, s, end, to, this, troop, come, thou, not, near, from, this, session, interdict, every, fowl, of, tyrant, wing, save, the, eagle, feather, d, king, keep, the, obsequy, so, strict, let, the, priest, in, surplice, white, that, defunctive, music, can, be, the, death, divining, swan, lest, the, requiem, lack, his, right, and, thou, treble, dated, crow, that, thy, sable, gender, mak, st, with, the, breath, thou, ...]","[let, bird, loudest, lay, sole, arabian, tree, herald, sad, trumpet, whose, sound, chaste, wings, obey, thou, shrieking, harbinger, foul, precurrer, fiend, augur, fever, end, troop, come, thou, near, session, interdict, every, fowl, tyrant, wing, save, eagle, feather, king, keep, obsequy, strict, let, priest, surplice, white, defunctive, music, death, divining, swan, lest, requiem, lack, right, thou, treble, dated, crow, thy, sable, gender, mak, st, breath, thou, giv, st, tak, st, mongst, mourners, shalt, thou, go, anthem, doth, commence, love, constancy, dead, phoenix, turtle, fled, 

* Tokenize (words with remove punctuation, all lowercase and  remove stop words, with speech_tagging)

In [47]:
pos_lambda = lambda x: nltk.pos_tag(x)
poem['tokens_pos'] = poem['tokens_no_stop'].map(pos_lambda)
poem.head(1)

,label,title,poem content,tokens_basic,tokens_clean,tokens_lower,tokens_no_stop,tokens_pos
0,sadness,Poem 1,"Let the bird of loudest lay\r\nOn the sole Arabian tree\r\nHerald sad and trumpet be,\r\nTo whose sound chaste wings obey.\r\n\r\nBut thou shrieking harbinger,\r\nFoul precurrer of the fiend,\r\nAugur of the fever's end,\r\nTo this troop come thou not near.\r\n\r\nFrom this session interdict\r\nEvery fowl of tyrant wing,\r\nSave the eagle, feather'd king;\r\nKeep the obsequy so strict.\r\n\r\nLet the priest in surplice white,\r\nThat defunctive music can,\r\nBe the death-divining swan,\r\nLest the requiem lack his right.\r\n\r\nAnd thou treble-dated crow,\r\nThat thy sable gender mak'st\r\nWith the breath thou giv'st and tak'st,\r\n'Mongst our mourners shalt thou go.\r\n\r\nHere the anthem doth commence:\r\nLove and constancy is dead;\r\nPhoenix and the Turtle fled\r\nIn a mutual flame from hence.\r\n\r\nSo they lov'd, as love in twain\r\nHad the essence but in one;\r\nTwo distincts, division none:\r\nNumber there in love was slain.\r\n\r\nHearts remote, yet not asunder;\r\nDistance and no space was seen\r\n'Twixt this Turtle and his queen:\r\nBut in them it were a wonder.\r\n\r\nSo between them love did shine\r\nThat the Turtle saw his right\r\nFlaming in the Phoenix' sight:\r\nEither was the other's mine.\r\n\r\nProperty was thus appalled\r\nThat the self was not the same;\r\nSingle nature's double name\r\nNeither two nor one was called.\r\n\r\nReason, in itself confounded,\r\n","[Let, the, bird, of, loudest, lay, On, the, sole, Arabian, tree, Herald, sad, and, trumpet, be, ,, To, whose, sound, chaste, wings, obey., But, thou, shrieking, harbinger, ,, Foul, precurrer, of, the, fiend, ,, Augur, of, the, fever, 's, end, ,, To, this, troop, come, thou, not, near., From, this, session, interdict, Every, fowl, of, tyrant, wing, ,, Save, the, eagle, ,, feather, 'd, king, ;, Keep, the, obsequy, so, strict., Let, the, priest, in, surplice, white, ,, That, defunctive, music, can, ,, Be, the, death-divining, swan, ,, Lest, the, requiem, lack, his, right., And, thou, treble-dated, crow, ,, That, ...]","[Let, the, bird, of, loudest, lay, On, the, sole, Arabian, tree, Herald, sad, and, trumpet, be, To, whose, sound, chaste, wings, obey, But, thou, shrieking, harbinger, Foul, precurrer, of, the, fiend, Augur, of, the, fever, s, end, To, this, troop, come, thou, not, near, From, this, session, interdict, Every, fowl, of, tyrant, wing, Save, the, eagle, feather, d, king, Keep, the, obsequy, so, strict, Let, the, priest, in, surplice, white, That, defunctive, music, can, Be, the, death, divining, swan, Lest, the, requiem, lack, his, right, And, thou, treble, dated, crow, That, thy, sable, gender, mak, st, With, the, breath, thou, ...]","[let, the, bird, of, loudest, lay, on, the, sole, arabian, tree, herald, sad, and, trumpet, be, to, whose, sound, chaste, wings, obey, but, thou, shrieking, harbinger, foul, precurrer, of, the, fiend, augur, of, the, fever, s, end, to, this, troop, come, thou, not, near, from, this, session, interdict, every, fowl, of, tyrant, wing, save, the, eagle, feather, d, king, keep, the, obsequy, so, strict, let, the, priest, in, surplice, white, that, defunctive, music, can, be, the, death, divining, swan, lest, the, requiem, lack, his, right, and, thou, treble, dated, crow, that, thy, sable, gender, mak, st, with, the, breath, thou, ...]","[let, bird, loudest, lay, sole, arabian, tree, herald, sad, trumpet, whose, sound, chaste, wings, obey, thou, shrieking, harbinger, foul, precurrer, fiend, augur, fever, end, troop, come, thou, near, session, interdict, every, fowl, tyrant, wing, save, eagle, feather, king, keep, obsequy, strict, let, priest, surplice, white, defunctive, music, death, divining, swan, lest, requiem, lack, right, thou, treble, dated, crow, thy, sable, gender, mak, st, breath, thou, giv, st, tak, st, mongst, mourners, shalt, thou, go, anthem, doth, commence, love, constancy, dead, phoenix, tur

* Compare the similar document using Count Vector and TF-IDF Vectorizer

In [48]:
# for this part I will take a class of poems and use those classes for the count vector and tf-idf vectorizer
# the sadness class is the one with the most amount of poems in the dataset, so I will be using that  class

sad_poems = poem[poem.label == "sadness" ]

sad_poems['label']

0      sadness
4      sadness
5      sadness
7      sadness
8      sadness
        ...   
432    sadness
436    sadness
439    sadness
441    sadness
449    sadness
Name: label, Length: 143, dtype: str

In [51]:
sad_poems['cleaned_content'] = sad_poems['tokens_no_stop'].apply(lambda tokens: " ".join(tokens))
corpus1 = sad_poems['cleaned_content']
corpus1.head()

0    let bird loudest lay sole arabian tree herald sad trumpet whose sound chaste wings obey thou shrieking harbinger foul precurrer fiend augur fever end troop come thou near session interdict every fowl tyrant wing save eagle feather king keep obsequy strict let priest surplice white defunctive music death divining swan lest requiem lack right thou treble dated crow thy sable gender mak st breath thou giv st tak st mongst mourners shalt thou go anthem doth commence love constancy dead phoenix turtle fled mutual flame hence lov love twain essence one two distincts division none number love slain hearts remote yet asunder distance space seen twixt turtle queen wonder love shine turtle saw right flaming phoenix sight either mine property thus appalled self single nature double name neither two one called reason confounded
4                                                                                                                                                                      

In [52]:

count_vect = CountVectorizer()

X_count = count_vect.fit_transform(sad_poems['cleaned_content'])

X_count_df = pd.DataFrame(X_count.toarray(), columns=count_vect.get_feature_names_out(), index=sad_poems['title'])
X_count_df.head()

,abhorring,abject,aboard,aboue,absence,absolute,absurd,abundance,abundant,abysm,...,youths,ypent,ys,ytorne,zar,zero,zeus,zones,zquez,zuan
title,,,,,,,,,,,,,,,,,,,,,
Poem 1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
Poem 5,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
Poem 6,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
Poem 8,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
Poem 9,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [53]:
Poem1 = list(X_count_df.loc['Poem 1'])
Poem1[:20]

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

In [56]:
from numpy import dot
from numpy.linalg import norm

cosine = lambda v1, v2: dot(v1, v2) / (norm(v1) * norm(v2))

In [ ]:
similarity = [cosine(Poem1, song) for song in X]
sorted(list(zip(similarity, beatles.song)), reverse=True)[1:6]

title,Poem 1,Poem 5,Poem 6,Poem 8,Poem 9,Poem 10,Poem 11,Poem 15,Poem 16,Poem 18,...,Poem 419,Poem 423,Poem 428,Poem 429,Poem 431,Poem 433,Poem 437,Poem 440,Poem 442,Poem 450
title,,,,,,,,,,,,,,,,,,,,,
Poem 1,1.000000,0.078878,0.091080,0.065142,0.290478,0.006115,0.102429,0.163943,0.174553,0.044097,...,0.208935,0.031353,0.024227,0.204797,0.018881,0.000000,0.101554,0.045255,0.022618,0.096838
Poem 5,0.078878,1.000000,0.066617,0.024957,0.082874,0.008200,0.072289,0.084055,0.046813,0.014783,...,0.154091,0.050450,0.043315,0.096116,0.037978,0.092555,0.090784,0.030342,0.026538,0.059024
Poem 6,0.091080,0.066617,1.000000,0.000000,0.038278,0.009469,0.033389,0.082126,0.015444,0.000000,...,0.145578,0.038837,0.062520,0.095130,0.000000,0.000000,0.087357,0.023357,0.021888,0.068155
Poem 8,0.065142,0.024957,0.000000,1.000000,0.037643,0.010642,0.107887,0.016782,0.000000,0.019185,...,0.009090,0.000000,0.007027,0.008910,0.049286,0.017159,0.009818,0.045939,0.044280,0.038300
Poem 9,0.290478,0.082874,0.038278,0.037643,1.000000,0.014135,0.037383,0.211766,0.317015,0.012741,...,0.048295,0.007247,0.037333,0.201187,0.000000,0.011396,0.052164,0.017434,0.006535,0.040698


In [59]:

from sklearn.feature_extraction.text import TfidfVectorizer

cv_tfidf = TfidfVectorizer(stop_words="english")
X_tfidf = cv_tfidf.fit_transform(sad_poems['cleaned_content']).toarray()
dt_tfidf = pd.DataFrame(X_tfidf, columns=cv_tfidf.get_feature_names_out())
dt_tfidf.index = ["Poem " + str(i+1) for i in range(len(sad_poems))]
dt_tfidf.head()

,abhorring,abject,aboard,aboue,absence,absolute,absurd,abundance,abundant,abysm,...,youths,ypent,ys,ytorne,zar,zero,zeus,zones,zquez,zuan
Poem 1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
Poem 2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
Poem 3,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
Poem 4,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
Poem 5,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [60]:
from numpy import dot
from numpy.linalg import norm

cosine = lambda v1, v2: dot(v1, v2) / (norm(v1) * norm(v2))

In [61]:
Poem1_vec = X_tfidf[0]
similarity_tfidf = [cosine(Poem1_vec, vec) for vec in X_tfidf]
sorted(list(zip(similarity_tfidf, dt_tfidf.index)), reverse=True)[1:6]

[(np.float64(0.2297260272581816), 'Poem 58'),
 (np.float64(0.16469850557692148), 'Poem 5'),
 (np.float64(0.15533984623004093), 'Poem 25'),
 (np.float64(0.14257074646092477), 'Poem 67'),
 (np.float64(0.13772709397301977), 'Poem 24')]

From the result above, it can be seen that Poem 1 is most similar to Poem 58 with a similarity score of 22.97%, indicating that they share some common words. Other poems with lower similarity scores suggest that even though they belong to the same class, their vocabulary and expressions differ more than the poem 58.

* Use the cosine similarity to measure the similarity of text

In [62]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity
corpus = sad_poems['cleaned_content'].tolist()
vectorizer = CountVectorizer()
X_count = vectorizer.fit_transform(corpus)
cos_sim_matrix = cosine_similarity(X_count)
cos_sim_df = pd.DataFrame(cos_sim_matrix, index=sad_poems['title'], columns=sad_poems['title'])
cos_sim_df.head()

title,Poem 1,Poem 5,Poem 6,Poem 8,Poem 9,Poem 10,Poem 11,Poem 15,Poem 16,Poem 18,...,Poem 419,Poem 423,Poem 428,Poem 429,Poem 431,Poem 433,Poem 437,Poem 440,Poem 442,Poem 450
title,,,,,,,,,,,,,,,,,,,,,
Poem 1,1.000000,0.078878,0.091080,0.065142,0.290478,0.006115,0.102429,0.163943,0.174553,0.044097,...,0.208935,0.031353,0.024227,0.204797,0.018881,0.000000,0.101554,0.045255,0.022618,0.096838
Poem 5,0.078878,1.000000,0.066617,0.024957,0.082874,0.008200,0.072289,0.084055,0.046813,0.014783,...,0.154091,0.050450,0.043315,0.096116,0.037978,0.092555,0.090784,0.030342,0.026538,0.059024
Poem 6,0.091080,0.066617,1.000000,0.000000,0.038278,0.009469,0.033389,0.082126,0.015444,0.000000,...,0.145578,0.038837,0.062520,0.095130,0.000000,0.000000,0.087357,0.023357,0.021888,0.068155
Poem 8,0.065142,0.024957,0.000000,1.000000,0.037643,0.010642,0.107887,0.016782,0.000000,0.019185,...,0.009090,0.000000,0.007027,0.008910,0.049286,0.017159,0.009818,0.045939,0.044280,0.038300
Poem 9,0.290478,0.082874,0.038278,0.037643,1.000000,0.014135,0.037383,0.211766,0.317015,0.012741,...,0.048295,0.007247,0.037333,0.201187,0.000000,0.011396,0.052164,0.017434,0.006535,0.040698


#### Conclusion

From this activity, I learned how to preprocess textual data by cleaning it, removing stopwords, and preparing it for analysis. I also learned how to represent text numerically using Count Vectorizer and TF-IDF Vectorizer, which allowed me to measure the similarity between poems. Using cosine similarity, I was able to identify which poems are most alike in terms of their content and vocabulary. Overall, I gained hands-on experience in text analysis and natural language processing, and I now understand how these techniques can be used to compare documents, explore patterns in writing, and find similarities in literary works. It was interesting to see how even poems within the same class can differ in expression yet still share measurable connections.